### PR TITLE
[CPU] Add data-tiling for s8s4s32 arm64 ukernels

### PR DESCRIPTION
**Pull request by mariecwhite**

### CODE REVIEW

**Comment by hanhanW** [comment]
Are we able to bring couple models with `i4` matmul to our benchmark suite? Otherwise, running benchmark seems not needed, and we might easily hit regression.

**Comment by mariecwhite** [comment]
> Are we able to bring couple models with `i4` matmul to our benchmark suite? Otherwise, running benchmark seems not needed, and we might easily hit regression.

We are planning to add Gemma int4 that Dan generated to the benchmarks. It requires changes to the infra to load weights at runtime. I can also add an i8i4 matmul mlir microbenchmark.

File: compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir [hunk 1]
```diff
@@ -1298,6 +1298,200 @@ func.func @matmul_lowering_i8i8i32_aarch64_i8mm() attributes {
 
 // -----
 
+#map = affine_map<(d0, d1, d2) -> (d0, d2)>
+#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
+#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
+func.func @matmul_lowering_i8i4i32_aarch64() attributes {
+  hal.executable.target = #hal.executable.target<"xyz", "xyz", {target_triple="aarch64-xyz-xyz"}>
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir:1305]: Is the intent to not have `ukernels = "all"` for this tests?

**Comment by mariecwhite** [comment]
> > Are we able to bring couple models with `i4` matmul to our benchmark suite? Otherwise, running benchmark seems not needed, and we might easily hit regression.
> 
> We are planning to add Gemma int4 that Dan generated to the benchmarks. It requires changes to the infra to load weights at runtime. I can also add an i8i4 matmul mlir microbenchmark.

Added microbenchmarks here: https://github.com/openxla/iree/pull/16839

File: compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir (hunk 1 above)
Comment (mariecwhite) [reply compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir:1305]: Correct, this tests the path where ukernels are not enabled so it'll fall into the `if (!hasUkernel(target)) { ... }` block. This checks that it'll use the 4x16x1 tile size inside of that if statement.

**Review by bjacob — APPROVED** [review]

**Review by dcaballe — APPROVED** [review]

**Pull request merged**